src/libs/elektra/kdb.c

```diff
@@ -362,591 +363,790 @@ static int ensureContract (KDB * handle, const KeySet * contract, Key * parentKe
+	Key * pluginsRoot = keyNew ("system:/", KEY_END);
+	ksRename (plugins, lookupHelper, pluginsRoot);
+	keyDel (pluginsRoot);
+
+	// find backend plugin
+	Key * backendPluginKey = ksLookupByName (plugins, "system:/backend", 0);
+	if (backendPluginKey == NULL)
 	{
-		splitDel (handle->split);
+		ELEKTRA_ADD_INSTALLATION_WARNINGF (errorKey, "The mountpoint '%s' defined in '%s' does not specify a backend plugin.",
+						   keyName (mountpoint), keyName (root));
+		keyDel (mountpoint);
+		keyDel (lookupHelper);
+		ksDel (plugins);
+		return false;
 	}
 
-	trieClose (handle->trie, errorKey);
+	// get definition section
+	keySetBaseName (lookupHelper, "definition");
+	KeySet * definition = ksBelow (elektraKs, lookupHelper);
+	Key * definitionRoot = keyNew ("system:/", KEY_END);
+	ksRename (definition, lookupHelper, definitionRoot);
+	keyDel (definitionRoot);
 
-	backendClose (handle->defaultBackend, errorKey);
-	handle->defaultBackend = 0;
+	keyDel (lookupHelper);
 
-	// not set in fallback mode, so lets check:
-	if (handle->initBackend)
+	// create mountpoint
+	if (keyGetNamespace (mountpoint) == KEY_NS_CASCADING)
 	{
-		backendClose (handle->initBackend, errorKey);
-		handle->initBackend = 0;
-	}
+		keySetNamespace (mountpoint, KEY_NS_SYSTEM);
+		if (!addDupMountpoint (mountpoints, mountpoint, plugins, definition, errorKey))
+		{
+			return false;
+		}
 
-	for (int i = 0; i < NR_GLOBAL_POSITIONS; ++i)
-	{
-		for (int j = 0; j < NR_GLOBAL_SUBPOSITIONS; ++j)
+		keySetNamespace (mountpoint, KEY_NS_USER);
+		if (!addDupMountpoint (mountpoints, mountpoint, plugins, definition, errorKey))
 		{
-			elektraPluginClose (handle->globalPlugins[i][j], errorKey);
+			return false;
 		}
-	}
 
-	if (handle->modules)
-	{
-		elektraModulesClose (handle->modules, errorKey);
-		ksDel (handle->modules);
+		keySetNamespace (mountpoint, KEY_NS_DIR);
+		if (!addDupMountpoint (mountpoints, mountpoint, plugins, definition, errorKey))
+		{
+			return false;
+		}
+
+		keySetNamespace (mountpoint, KEY_NS_PROC);
+		if (!addDupMountpoint (mountpoints, mountpoint, plugins, definition, errorKey))
+		{
+			return false;
+		}
+
+		keyDel (mountpoint);
 	}
 	else
 	{
-		ELEKTRA_ADD_RESOURCE_WARNING (errorKey, "Could not close modules: modules were not open");
+		Plugin * backendPlugin = *(Plugin **) keyValue (backendPluginKey);
+		addMountpoint (mountpoints, mountpoint, backendPlugin, plugins, definition);
 	}
 
-	if (handle->global) ksDel (handle->global);
-
-	elektraFree (handle);
-
-	keySetName (errorKey, keyName (initialParent));
-	keySetString (errorKey, keyString (initialParent));
-	keyDel (initialParent);
-	errno = errnosave;
-	return 0;
+	return true;
 }
 
-/**
- * @internal
- *
- * @brief Check if an update is needed at all
- *
- * @retval -2 cache hit
- * @retval -1 an error occurred
- * @retval 0 no update needed
- * @retval number of plugins which need update
- */
-static int elektraGetCheckUpdateNeeded (Split * split, Key * parentKey)
+// FIXME [new_backend]: write tests
```

Comment:
This doesn't make much sense here unless it is a marker for yourself.

Reply:
Why? We should have tests for `elektraMountpointsParse` that's why I put a marker here

Reply:
> If there are no tests, it is shown by code coverage.

No, this won't show up. This function is called by `kdbOpen`. So any test that uses the KDB "covers" this function. But the errors in those tests will not at all indicate that there is a problem in `elektraMountpointsParse` and there will be even less indication of what the actual problem is.

So to be more specific, the function is tested, but doesn't have unit tests.

Reply:
Ok, then please add to doc/todo/TESTING